annotation-processor/src/main/java/org/hibernate/validator/ap/classchecks/AbstractMethodOverrideCheck.java

```diff
@@ -0,0 +1,338 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.ap.classchecks;
+
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.Name;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.util.Elements;
+import javax.lang.model.util.Types;
+
+import org.hibernate.validator.ap.checks.ConstraintCheckIssue;
+import org.hibernate.validator.ap.util.CollectionHelper;
+
+/**
+ * Abstract base class for {@link ClassCheck} implementations that check overridden methods.
+ *
+ * @author Marko Bekhta
+ */
+public abstract class AbstractMethodOverrideCheck extends AbstractClassCheck {
+
+	private static final String JAVA_LANG_OBJECT = "java.lang.Object";
+
+	private static final String JAVAX_VALIDATION_CONSTRAINT = "javax.validation.Constraint";
+
+	private final Elements elementUtils;
+
+	private final Types typeUtils;
+
+	public AbstractMethodOverrideCheck(Elements elementUtils, Types typeUtils) {
+		this.elementUtils = elementUtils;
+		this.typeUtils = typeUtils;
+	}
+
+	@Override
+	public Collection<ConstraintCheckIssue> checkMethod(ExecutableElement currentMethod) {
+		if ( !needToPerformAnyChecks( currentMethod ) ) {
+			return Collections.emptySet();
+		}
+
+		TypeElement currentTypeElement = getEnclosingTypeElement( currentMethod );
+
+		// if current type is java.lang.Object then we can continue without doing any other checks
+		if ( isJavaLangObjectOrNull( currentTypeElement ) ) {
+			return Collections.emptySet();
+		}
+		// find if there's a method that was overridden by the current one.
+		Map<Boolean, List<ExecutableElement>> overriddenMethods = findAllOverriddenElements( currentTypeElement, currentMethod );
+		if ( overriddenMethods.get( Boolean.TRUE ).isEmpty() && overriddenMethods.get( Boolean.FALSE ).isEmpty() ) {
+			return Collections.emptySet();
+		}
+
+		// if there's more than one overridden originally declared method we need to make sure that all of them match
+		Set<ConstraintCheckIssue> errors = CollectionHelper.newHashSet();
+		for ( ExecutableElement firstExecutable : overriddenMethods.get( Boolean.TRUE ) ) {
```

Comment:
OK, so I did my homework (finally, you might say) and the rules for parameters are much stricter than what you implemented. I'm a bit surprised you followed this road, we probably didn't guide you very well.

So the rules for parameters constraints are pretty clear:
* you can't define a constraint on a parameter of an overriding/implementing method - so it's not a matter of the constraints match or whatever: you simply can't do it. So you need to check that methods having constraint annotations in the hierarchy does not override another method. I think it's far more simple that what you did.
* the second rule is that if you have 2 parallel hierarchies both of which implementing the same method, you can't define a parameter constraint at all for this method (anywhere in the hierarchy, not even once). I think this check should be done first.

Tell me if I'm wrong but I think we were a bit off before.

You also have the rules for the `@Valid` annotation that might be interesting to implement (referred as cascaded in the doc links below).

See https://docs.jboss.org/hibernate/stable/validator/reference/en-US/html_single/#section-method-constraints-inheritance-hierarchies for reference on the constraint parameters. http://beanvalidation.org/latest-draft/spec/#constraintdeclarationvalidationprocess-methodlevelconstraints-inheritance is also interesting to understand the whole thing.

Another thing: we need to be very careful on the error message we provide: the user should have all the required information to fix the issue. This is especially true when the errors are not about the class itself but its hierarchy (we need both classes/interfaces in the error message).